Use std::align_alloc in file_data_loader

lucylq opened the pull request:
Summary:
Issue with aligned buffers: P1800967583

The alignment requested is 16, and std::max_align_t is also 16. This means we do not need to pad the size to meet any alignment.

However, the buffer we get from malloc is aligned to 8, not 16. When we try to align the buffer to 16, we overflow the original buffer (as it wasn't padded) and error out.

Seems like malloc is not guaranteed to return 8 or 16 byte-aligned buffers, so also a bit hard to test definitively. So far we've only seen this when the buffer size is small (size 2, 4)
```
The malloc(), calloc(), realloc(), and reallocarray() functions
return a pointer to the allocated memory, which is suitably
aligned for any type that fits into the requested size or less.
```

Use std::aligned_alloc (C++17) to ensure buffer is aligned.

Differential Revision: D74041198

larryliu0820 reviewed (approved):
Is it possible to add some tests?

lucylq commented:
> Is it possible to add some tests?

alignment is covered by the file_data_loader_tests: https://github.com/pytorch/executorch/blob/3ca6cbbff564d52bb5c206539d25fadb5f1eb854/extension/data_loader/test/file_data_loader_test.cpp#L68

I think we shouldn't have this error now that we've moved to aligned_alloc (hopefully), probably the main thing is to make sure oss ci passes on macos.

Are you thinking of a different test though?

lucylq committed:
Introduce ET_ALIGN_ALLOC for portable aligned alloc (#10660)

Summary:

Issue with aligned buffers: P1800967583

The alignment requested is 16, and std::max_align_t is also 16. This means we do not need to pad the size to meet any alignment.

However, the buffer we get from malloc is aligned to 8, not 16. When we try to align the buffer, we overflow and error out.

Seems like malloc is not guaranteed to return 8 or 16 byte-aligned buffers, so also a bit hard to test definitively. So far we've only seen this when the buffer size is small (size 2, 4)
```
The malloc(), calloc(), realloc(), and reallocarray() functions
return a pointer to the allocated memory, which is suitably
aligned for any type that fits into the requested size or less.
```

Use std::aligned_alloc (C++17) to ensure buffer is aligned.

For systems that do not have aligned_alloc (or posix_memalign) fallback to malloc.

The malloc implementation is similar to what file_data_loader.cpp does. Except, we do not have a custom free function from FreeableBuffer, so we store an offset just before the aligned ptr to free the actual buffer.

1. Allocate via malloc; buffer = malloc(size + sizeof(uint16_t) + alignment - 1)
- size: the size requested.
- sizeof(uint16_t): a place to store the offset between the aligned buffer and the original ptr.
- alignment-1: extra padding to allow for alignment.
2. Align (buffer + sizeof(uint16_t)) to alignment. This (usually) pushes the buffer forward by `alignment`.
3. Store the difference of (aligned_ptr - buffer).

The memory will look like this:
| buffer start | maybe padding | offset (aligned_buffer - buffer) | aligned_buffer start | maybe padding | buffer end |

We should have between offset_size (2) and 1 aligned block before the actual aligned buffer.

https://embeddedartistry.com/blog/2017/02/22/generating-aligned-memory/

NOTE: this increase binary size (linux, clang) by 8 bytes, so also raising it there.

Reviewed By: larryliu0820, mcr229

Differential Revision: D74041198

swolchok commented:
> Seems like malloc is not guaranteed to return 8 or 16 byte-aligned buffers

it is [*supposed* to be guaranteed](https://en.cppreference.com/w/cpp/memory/c/malloc), but [reality does not always conform to standards](https://github.com/jemalloc/jemalloc/issues/1533#issuecomment-507915829) :\

swolchok commented on runtime/platform/compiler.h:
```diff
@@ -171,6 +171,101 @@
+#define ET_ALIGNED_ALLOC(alignment, size) \
+  _aligned_malloc(((size + alignment - 1) & ~(alignment - 1)), (alignment))
+#define ET_ALIGNED_FREE(ptr) _aligned_free(ptr)
+#elif defined(__APPLE__)
+#include <stdlib.h> // For posix_memalign and free
+inline void* et_apple_aligned_alloc(size_t alignment, size_t size) {
+  void* ptr = nullptr;
+  // The address of the allocated memory must be a multiple of sizeof(void*).
+  if (alignment < sizeof(void*)) {
+    alignment = sizeof(void*);
+  }
+  if (posix_memalign(
+          &ptr, alignment, (size + alignment - 1) & ~(alignment - 1)) != 0) {
+    return nullptr;
+  }
+  return ptr;
+}
```
would be nice to include citations for why we need all these varied solutions despite std::aligned_alloc being required in C++17.  https://gitlab.com/gromacs/gromacs/-/issues/3968 is one example but I'm curious where you got this from.

lucylq replied:
Ah from: https://www.internalfb.com/code/fbsource/[0afaf0bdc6ee692ec9a9b3e3c57bca4af2930abc]/third-party/pocket_fft/pocketfft_hdronly.h?lines=156%2C165

And mostly from ci tests - xtensa seems to only have malloc available? Thanks for the link!
